velox/functions/sparksql/URLFunctions.h

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <boost/regex.hpp>
+#include <folly/CPortability.h>
+#include <cctype>
+#include "velox/functions/Macros.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+/// Performs initial validation of the URI.
+/// Checks if the URI contains ascii whitespaces or
+/// unescaped '%' chars.
+bool isValidURI(StringView input) {
+  const char* p = input.data();
+  const char* end = p + input.size();
+  char buf[3];
+  buf[2] = '\0';
+  char* endptr;
+  for (; p < end; ++p) {
+    if (facebook::velox::functions::stringImpl::isAsciiWhiteSpace(*p)) {
+      return false;
+    }
+
+    if (*p == '%') {
+      if (p + 2 < end) {
+        buf[0] = p[1];
+        buf[1] = p[2];
+        strtol(buf, &endptr, 16);
+        p += 2;
+        if (endptr != buf + 2) {
+          return false;
+        }
+      } else {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+static const boost::regex kUriRegex(
+    "(([a-zA-Z][a-zA-Z0-9+.-]*):)?" // scheme:
+    "([^?#]*)" // authority and path
+    "(?:\\?([^#]*))?" // ?query
+    "(?:#(.*))?"); // #fragment
+template <typename T>
+T submatch(const boost::cmatch& match, int indx);
+
+template <>
+FOLLY_ALWAYS_INLINE std::string_view submatch(
+    const boost::cmatch& match,
+    int idx) {
+  const auto& sub = match[idx];
+  return std::string_view(sub.first, sub.length());
+}
+
+template <>
+FOLLY_ALWAYS_INLINE StringView submatch(const boost::cmatch& match, int idx) {
```

Comment:
Would it make sense to for all 'submatch' functions to use std::string_view for return type?

Reply:
One reason this function is templated is found in the 3 parameter call function, which extracts the value of a key from a query. 

Below is that function, but it is modified with the detail namespace and a version of submatch that only returns std::string_view. We compare an `arg_type<Varchar>& key` (velox::StringView) to a `std::string_view k`

```cpp

    auto query = detail::submatch(match, kQuery);
    const boost::cregex_iterator begin(
        query.data(), query.data() + query.size(), kQueryParamRegex);
    boost::cregex_iterator end;

    for (auto it = begin; it != end; ++it) {
      if (it->length(2) != 0) { // key shouldn't be empty.
        auto k = detail::submatch((*it), 2);
        if (key.compare(k) == 0) {
          auto value = detail::submatch((*it), 3);
          if (value != "") {
            result.setNoCopy(StringView(value.data(), value.size()));
            return true;
          } else {
            return false;
          }
        }
      }
    }
```

In this version, we run into incompatibilities between std::string_view and StringView. First, key.compare(k) fails to compile because velox::StringView does not support comparisons against std::string_views. Second, result.setNoCopy does not support std::string_views, so we are forced to convert to StringView.

In this case it is thus easier to start with a StringView since we do not need a std::string_view.

A second reason, that will be included in the changes from this review, is when we know that we won't need to merge any StringViews (anytime simpleMatch is called), it is also better to avoid converting to a velox::StringView in the result.setNoCopy() call.